### PR TITLE
CgroupsV2 based Flatcar version support

### LINKED ISSUE #35: Add support for newer versions of Flatcar using cgroup v2 and systemd cgroup driver
**How to categorize this issue?**

/area os
/kind enhancement
/priority 3

**What would you like to be added**:
Add support for newer versions of Flatcar using cgroup v2 and systemd cgroup driver

1. Kubernetes <=v1.18 does not support cgroup v2, needs to be decided if this extension will support older k8s versions with newer Flatcar versions. The good thing is that k8s 1.18 is out of maintenance since long time and it is highly unlikely to be used.
1. cgroup v2 is not working with `cgroupfs` driver which is explicitly set in the [kubelet config](https://github.com/gardener/gardener/blob/bfd8cef1f1fe920b18dd65ffde7523702f9d7db9/pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/kubelet/config.go#L54). This extension needs to change it to `systemd`

More details what needs to be changed can be found in [this](https://www.flatcar-linux.org/docs/latest/container-runtimes/switching-to-unified-cgroups/#kubernetes) nice blog from Kinvolk.

**Why is this needed**:
To enable support for newer flatcar versions with this extension.

### PR DESCRIPTION
**Pull request by lukasgr90**
**How to categorize this PR?**

/area os
/kind enhancement

**What this PR does / why we need it**:

An additional oneshot systemd unit now patches the kubelet configuration to use the systemd cgroupfs driver, if cgroupsv2 is enabled. This is required to start the kubelet on recent Flatcar versions, which by default boot with cgroupsv2 enabled.

**Which issue(s) this PR fixes**:
Fixes #35 

**Special notes for your reviewer**:

**Release note**:

```other operator
Added support for Flatcar versions, which enable cgroupsv2 by default (2969.0.0 and newer).
```

### CODE REVIEW

**Comment by gardener-robot-ci-2** [comment]
Thank you @lukasgr90 for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

File: pkg/coreos/actuator_reconcile.go [hunk 1]
```diff
@@ -189,3 +198,36 @@ func isContainerdEnabled(criConfig *extensionsv1alpha1.CRIConfig) bool {
...
+	var additionalUnitNames []string
+
+	cloudConfig.CoreOS.Units = append(
+		cloudConfig.CoreOS.Units,
+		Unit{
+			Name:    "enable-cgroupsv2.service",
+			Command: "start",
+			Enable:  true,
+			Content: `[Unit]
+Description=Oneshot unit used to patch the kubeconfig for cgroupsv2.
```
Comment (rfranzke) [inline pkg/coreos/actuator_reconcile.go:212]:
```suggestion
Description=Oneshot unit used to patch the kubelet config for cgroupsv2.
```

**Review by rfranzke — COMMENTED** [review]
Would it also be possible to implement it via the `controlplane` webhook (https://github.com/gardener/gardener-extension-provider-aws/blob/14199bdc178b1fb0403e684fde64162e40c3c7d2/pkg/webhook/controlplane/ensurer.go#L431) or does it need to discover the `/sys/fs/cgroup/cgroup.controllers` file on the nodes?

File: pkg/coreos/actuator_reconcile.go (hunk 1 above)
Comment (lukasgr90) [reply pkg/coreos/actuator_reconcile.go:212]: If the patch is required or not, depends basically on the version of the Flatcar image.

If I see correctly, the chosen machineImage (=> version) is available in the controlplane webhook (as part of the Cluster definition). Thus it should be possible to implement it also there.

**Comment by MrBatschner** [comment]
@rfranzke We have the same issue on Garden Linux. We are planning to release future Garden Linux releases with cgroup v2 enabled and thus need to configure the kubelet accordingly. We could ask the control plane, what GL version the worker has and base the configuration on that but that would - in worst case - require to maintain a list of Garden Linux versions and their default cgroup setting. Determining from the running system which cgroup version is active and reconfigure the kubelet before it comes up seems a lot more straightforward and less error-prone, don't you think?

**Comment by rfranzke** [comment]
> Determining from the running system which cgroup version is active and reconfiguring the kubelet before it comes up seems a lot more straightforward and less error-prone, don't you think?

Yes, sure, it sounds more sustainable. I was just asking whether we have other options, but it seems the proposed option is the best for now (in the future, I expect cgroups-v1 to disappear entirely anyways, right? Then we don't need such code anymore and can build on the assumption that cgroups-v2 are used).

**Review by MrBatschner — CHANGES_REQUESTED** [review]
see nits above

But there is more. I tested the changes with with FlatCar `3139.2.0`, see Shoot manifest below.

<details>

```yaml
kind: Shoot
apiVersion: core.gardener.cloud/v1beta1
metadata:
  name: aws-tst03
  annotations:
    confirmation.gardener.cloud/deletion: "true"
spec:
  cloudProfileName: aws
  kubernetes:
    allowPrivilegedContainers: true
    kubeAPIServer:
      enableBasicAuthentication: false
      requests:
        maxNonMutatingInflight: 400
        maxMutatingInflight: 200
      enableAnonymousAuthentication: false
      eventTTL: 1h0m0s
    kubeControllerManager:
      nodeCIDRMaskSize: 24
      podEvictionTimeout: 2m0s
      nodeMonitorGracePeriod: 2m0s
    kubeProxy:
      mode: IPTables
      enabled: true
    kubelet:
      failSwapOn: true
      kubeReserved:
        cpu: 80m
        memory: 1Gi
        pid: 20k
      imageGCHighThresholdPercent: 50
      imageGCLowThresholdPercent: 40
      serializeImagePulls: true
    version: 1.22.7
    verticalPodAutoscaler:
      enabled: true
      evictAfterOOMThreshold: 10m0s
      evictionRateBurst: 1
      evictionRateLimit: -1
      evictionTolerance: 0.5
      recommendationMarginFraction: 0.15
      updaterInterval: 1m0s
      recommenderInterval: 1m0s
    enableStaticTokenKubeconfig: true
  networking:
    type: calico
    nodes: 10.248.0.0/16
    pods: 10.96.0.0/16
    services: 10.112.0.0/16
  maintenance:
    autoUpdate:
      kubernetesVersion: true
      machineImageVersion: true
    timeWindow:
      begin: 030000+0200
      end: 040000+0200
  provider:
    type: aws
    controlPlaneConfig:
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      kind: ControlPlaneConfig
    infrastructureConfig:
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      kind: InfrastructureConfig
      networks:
        vpc:
          cidr: 10.248.0.0/16
        zones:
          - name: eu-west-1c
            workers: 10.248.0.0/18
            internal: 10.248.64.0/18
            public: 10.248.128.0/18
    workers:
      - cri:
          name: containerd
        name: worker-old
        machine:
          type: m5.large
          image:
            name: coreos
            version: 3139.2.0
        maximum: 2
        minimum: 1
        maxSurge: 1
        maxUnavailable: 0
        volume:
          type: gp2
          size: 50Gi
        zones:
          - eu-west-1c
        systemComponents:
          allow: true
      - cri:
          name: containerd
        name: worker-new
        machine:
          type: m5.large
          image:
            name: coreos
            version: 3139.2.0
        maximum: 2
        minimum: 1
        maxSurge: 1
        maxUnavailable: 0
        volume:
          type: gp2
          size: 50Gi
        zones:
          - eu-west-1c
        systemComponents:
          allow: true
  purpose: evaluation
  region: eu-west-1
  secretBindingName: <redacted>
  seedName: dev-seed-aws
  systemComponents:
    coreDNS:
      autoscaling:
        mode: horizontal
```

</details>


All Pods go into `CrashLoopBackoff`.

```
$ k describe po test-pod
[...]
Events:
  Type     Reason          Age                     From               Message
  ----     ------          ----                    ----               -------
  Normal   Scheduled       8m53s                   default-scheduler  Successfully assigned default/tbsuse-debug-6xn48 to ip-10-248-60-187.eu-west-1.compute.internal
  Normal   Pulled          8m53s                   kubelet            Successfully pulled image "thbb/tbsuse:15" in 690.226441ms
  Normal   Pulled          8m50s                   kubelet            Successfully pulled image "thbb/tbsuse:15" in 702.452842ms
  Normal   Pulling         8m31s (x3 over 8m54s)   kubelet            Pulling image "thbb/tbsuse:15"
  Normal   Created         8m30s (x3 over 8m53s)   kubelet            Created container tbsuse
  Normal   Started         8m30s (x3 over 8m53s)   kubelet            Started container tbsuse
  Normal   Pulled          8m30s                   kubelet            Successfully pulled image "thbb/tbsuse:15" in 681.561026ms
  Normal   Killing         7m2s (x3 over 8m52s)    kubelet            Stopping container tbsuse
  Normal   SandboxChanged  7m2s (x3 over 8m51s)    kubelet            Pod sandbox changed, it will be killed and re-created.
  Warning  BackOff         3m53s (x18 over 8m48s)  kubelet            Back-off restarting failed container
```

The extension works as expected, I can see that the script gets deployed to `/opt/bin/configure-cgroupsv2.sh`, the systemd unit runs and `/var/lib/kubelet/config/kubelet` has the systemd cgroup driver set.

However, containerd still seems to resort to the cgroupfs driver as a quick check reveals:

```
$ cat /host/etc/containerd/config.toml
[...]
          [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc.options]
[...]
            SystemdCgroup = false
[...]
```

This discrepancy might be the reason for all Pods getting stuck in a CrashLoop so please check what else needs to be done to get happy containers on a cgroup v2 enabled system.

/do-not-merge

**Comment by lukasgr90** [comment]
Thank you very much for the reviews so far! :-)

Shame on me that I missed the fault with the containerd. It should be fixed now. No pod restarts in kube-system (of my test containerd / Flatcar 3139.2.0 shoot) so far / the last 20 minutes. But I will do more testing tomorrow.

Also tomorrow I will work on the open naming issues.

**Comment by MrBatschner** [comment]
> Shame on me that I missed the fault with the containerd. It should be fixed now. No pod restarts in kube-system (of my test containerd / Flatcar 3139.2.0 shoot) so far / the last 20 minutes.

Yep, same here, had a Shoot running with your latest changes and it looked quite healthy for the better part of a day. So that looks good now.

/unhold

**Review by MrBatschner — APPROVED** [review]
/lgtm

**Review by rfranzke — APPROVED** [review]

**Review by vpnachev — APPROVED** [review]
/lgtm

**Comment by christianhuening** [comment]
@vpnachev @rfranzke @MrBatschner can this be merged and released now? or is there a cadence you're following for new releases? We would be keen to have these changes by @lukasgr90 so we can upgrade flatcar.

**Comment by rfranzke** [comment]
@MrBatschner could you take care of the release?

**Comment by MrBatschner** [comment]
> @MrBatschner could you take care of the release?

[Done.](https://github.com/gardener/gardener-extension-os-coreos/releases/tag/v1.11.0)

**Pull request merged**